File: benchmarks/c/mixed_pipeline.c
#include <stdint.h>
#include <stdio.h>

typedef struct {
    int64_t count;
    int64_t total;
} Stat;

int main(void) {
    Stat stats[9] = {{0, 0}};
    int64_t total = 0;

    for (int i = 0; i < 5001; ++i) {
        int key = i % 9;
        int value = 10 + (i % 23);
        stats[key].count += 1;
        stats[key].total += value;
    }

    for (int i = 0; i < 9; ++i) {
        total += stats[i].count * stats[i].total;
    }

    printf("%lld\n", (long long) total);
    return 0;
}

File: benchmarks/python/mixed_pipeline.py
from dataclasses import dataclass


@dataclass(slots=True)
class Stat:
    count: int
    total: int


def make_rows(repeats: int) -> list[str]:
    return [f"cat{i % 9},{10 + (i % 23)}" for i in range(repeats)]


def main() -> None:
    rows = make_rows(5001)
    stats: dict[str, Stat] = {}

    for row in rows:
        key, value_text = row.split(",")
        value = int(value_text)
        current = stats.get(key, Stat(0, 0))
        stats[key] = Stat(current.count + 1, current.total + value)

    total = 0
    for i in range(9):
        current = stats[f"cat{i}"]
        total += current.count * current.total

    print(total)


if __name__ == "__main__":
    main()
